builtins: Implement unaccent function

mknycha opened the pull request:
Unaccent function should work the same as in PostgreSQL.

Release note (sql change): Implement the string builtin `unaccent`.

Described in #41299

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=54628) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=54628) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=54628) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54628)
<!-- Reviewable:end -->

mknycha commented on pkg/sql/logictest/testdata/logic_test/builtin_function:
```diff
@@ -102,6 +102,26 @@ SELECT upper('roacH7')
+hELLO`
+
+query T
+SELECT unaccent('some­thing')
```
There is actually `U+00AD` in the middle, it's visible in git diff.

otan replied:
we can combine these into one test case that's easier to mess with if you do:
```
SELECT unaccent(str) FROM ( VALUES 
   ('no_special_CHARACTERS1!'),
   ('Żółć')
) tbl(str)
```

mknycha replied:
Ok, done

mknycha commented:
I forgot to add a label "first PR", sorry

otan commented on pkg/sql/sem/builtins/unaccent_dictionary.go:
```diff
@@ -0,0 +1,1519 @@
+// Copyright 2015 The Cockroach Authors.
```
nit: 2020.

I think this may be useful as a separate package, i.e. in `pkg/util/unaccent/unaccent.go`

mknycha replied:
Ok, done

otan commented on pkg/sql/sem/builtins/unaccent_dictionary.go:
```diff
@@ -0,0 +1,1519 @@
+
+package builtins
+
+var unaccentDictionary = map[string]string{
```
i'm curious - should this be `map[rune]string`?

otan replied:
(if not, I'm not sure how `strings.Split(s, "")` above works with this dictionary.

mknycha replied:
You're right, I have just realized that `strings.Split` may not work correctly. I guess that `map[rune]string` is indeed a way to go. Some "characters" in the dictionary consist of two code points, so I need to modify this dictionary a bit.

otan commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -260,6 +260,24 @@ var builtins = map[string]builtinDefinition{
+	"unaccent": makeBuiltin(tree.FunctionProperties{Category: categoryString},
+		stringOverload1(
+			func(evalCtx *tree.EvalContext, s string) (tree.Datum, error) {
+				separatedStrings := strings.Split(s, "")
```
nit: this is more efficient if we use a `strings.Builder` to write out the whole string.
I feel as if we should be iterating rune by rune, something like

```
var b strings.Builder
var changed bool
for _, ch := range s {
   if repl, ok := unaccent.Characters[ch]; ok {
      b.WriteString(repl)
      changed = true
   } else {
      b.WriteRune(ch)
   }
}

ret := s
// .String() can be expensive
if changed {
   ret = b.String()
}
return tree.NewDString(ret), nil
```

mknycha replied:
Ok, done

otan reviewed (commented):
hello @mknycha, thanks for doing this!
I was away last week, so apologies for the delay.
I have a couple of small comments.

mknycha committed:
builtins: Implement unaccent function

Unaccent function should work the same as in PostgreSQL.

Release note (sql change): Implement the string builtin unaccent.

otan commented:
looks like some linting issues -- i'll merge it after those are fixed!

otan commented:
bors r+

thank you for the contribution!
